src/ngx_http_lua_ssl_session_fetchby.c

```diff
@@ -376,10 +361,124 @@ ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
         ngx_http_lua_close_fake_connection(fc);
     }
 
+    return res;
+}
+
+
+#ifdef HAVE_SSL_CLIENT_HELLO_CB_SUPPORT
+int
+ngx_http_lua_ssl_client_hello_handler(ngx_ssl_conn_t *ssl_conn,
+    int *al, void *arg)
+{
+    int                len;
+    ngx_int_t          rc;
+    const u_char      *id;
+
+    len = SSL_client_hello_get0_session_id(ssl_conn, &id);
+
+    if (len <= 0) {
+        return SSL_CLIENT_HELLO_SUCCESS;
+    }
+
+    rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, id, len);
+
+    if (rc == NGX_AGAIN) {
+        return SSL_CLIENT_HELLO_RETRY;
+    }
+
+    return SSL_CLIENT_HELLO_SUCCESS;
+}
+
+
+ngx_ssl_session_t *
+ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
+    const u_char *id, int len, int *copy)
+{
+    ngx_connection_t                *c;
+    ngx_http_lua_ssl_ctx_t          *cctx;
+
+    /* set copy to 0 as we expect OpenSSL to handle
+     * the memory of returned session */
+
+    *copy = 0;
+
+    c = ngx_ssl_get_connection(ssl_conn);
+
+    cctx = ngx_http_lua_ssl_get_ctx(c->ssl->connection);
+
+    if (cctx && cctx->done) {
+        return cctx->session;
+    }
+
     return NULL;
 }
 
 
+#else
+
+/* cached session fetching callback to be set with SSL_CTX_sess_set_get_cb */
+ngx_ssl_session_t *
+ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
+#if OPENSSL_VERSION_NUMBER >= 0x10100003L
+    const
+#endif
+    u_char *id, int len, int *copy)
+{
+    ngx_int_t                        rc;
+    ngx_connection_t                *c;
+    ngx_http_lua_ssl_ctx_t          *cctx;
+
+    /* set copy to 0 as we expect OpenSSL to handle
+     * the memory of returned session */
+
+    *copy = 0;
+
+    c = ngx_ssl_get_connection(ssl_conn);
+
+    rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, id, len);
+
+    if (rc == NGX_AGAIN) {
+
+#ifndef SSL_ERROR_PENDING_SESSION
+
+        return SSL_magic_pending_session_ptr();
```

Comment:
When compiling this patch _without_ the nginx `ssl_sess_cb_yield` patch, one now gets this warning:
```
gcc -c -I/home/chasum/code/openresty/openresty-dev/ngx_lua/deps/luajit2-v2.1-agentzh/include//luajit-2.1  -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O0 -g -DNGXs/include -I/home/chasum/code/openresty/openresty-dev/ngx_lua/deps/include -DNDK_SET_VAR -DNDhome/chasum/code/openresty/openresty-dev/ngx_lua/ngx_devel_kit/objs -I objs/addon/ndk -I /homresty/openresty-dev/ngx_lua/deps/include/libdrizzle-1.0 -I /home/chasum/code/openresty/openreresty-dev/ngx_lua/stream-lua-nginx-module/src/api -I objs -I src/http -I src/http/modules -I um/code/openresty/openresty-dev/ngx_lua/ngx_devel_kit/src -I /home/chasum/code/openresty/opendev/ngx_lua/deps/luajit2-v2.1-agentzh/include/luajit-2.1 -I src/stream \
        -o objs/addon/src/ngx_http_headers_more_util.o \
        /home/chasum/code/openresty/openresty-dev/ngx_lua/headers-more-nginx-module/src/ngx_h
/home/chasum/code/openresty/openresty-dev/ngx_lua/lua-nginx-module/src/ngx_http_lua_ssl_sessi
/home/chasum/code/openresty/openresty-dev/ngx_lua/lua-nginx-module/src/ngx_http_lua_ssl_sessid you mean ‘SSL_has_matching_session_id’? [-Werror=implicit-function-declaration]
         return SSL_magic_pending_session_ptr();
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                SSL_has_matching_session_id
/home/chasum/code/openresty/openresty-dev/ngx_lua/lua-nginx-module/src/ngx_http_lua_ssl_sessia ‘struct ssl_session_st *’} makes pointer from integer without a cast [-Werror=int-conversio
         return SSL_magic_pending_session_ptr();
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

Seems like this branch needs to be better guarded.